File: src/player.py
from typing import List, Optional

from entity import Entity
from item import Item
from room import Room


class Player(Entity):
    _current_room: Room
    _items: List[Item]

    def __init__(self, name: str, room: Room):
        super().__init__(name, "PC")
        self._current_room = room
        self._items = []

    def __str__(self):
        details = f'==={self.name}==='
        for item in self._items:
            details += str(item)
        details += '\n\n'
        return details

    def look(self):
        print(self._current_room)

    def move(self, direction: str):
        if direction is 'n':
            direction = 'north'

        if direction is 's':
            direction = 'south'

        if direction is 'e':
            direction = 'east'

        if direction is 'w':
            direction = 'west'

        room = getattr(self._current_room, direction)

        if not room():
            raise TypeError

        self._current_room = room()

    def inventory(self):
        print(self)
        input('Press ENTER key to continue...')

    def take(self, item_name: str):
        items = self._current_room.items()
        item = next((item for item in items if item.has_name(item_name)))
        self._current_room.take(item)
        self._items.append(item)
        item.on_take()

    def drop(self, item_name: str):
        items = self._items
        item = next((item for item in items if item.has_name(item_name)))
        self._items.remove(item)
        self._current_room.drop(item)
        item.on_drop()

    def room(self):
        return self._current_room

    def items(self):
        return self._items

File: src/player.ts
import { Entity } from './entity';
import { Item } from './item';
import { Room } from './room';

export class Player extends Entity {
  private currentRoom: Room;
  private items: Item[];

  constructor(name: string, room: Room) {
    super(name, 'PC');
    this.currentRoom = room;
    this.items = [];
  }

  public toString() {
    let details = `===${this.name}===\n`;

    for (const item of this.items) {
      details += item.toString();
    }

    details += '\n';
    return details;
  }

  public look() {
    return this.currentRoom;
  }

  public move(room: Room) {
    this.currentRoom = room;
  }

  public take(itemName: string) {
    const item = this.currentRoom.items.find((item) => item.hasName(itemName));
    if (item === undefined) throw new Error('INVALID ITEM!');
    this.currentRoom.take(item);
    this.items.push(item);
    item.onTake();
  }

  public drop(itemName: string) {
    const index = this.items.findIndex((item) => item.hasName(itemName));
    if (index === -1) throw new Error('INVALID ITEM!');
    const [item] = this.items.splice(index, 1);
    this.currentRoom.drop(item);
    item.onDrop();
  }

  public lookNorth() {
    return this.currentRoom.north;
  }

  public lookSouth() {
    return this.currentRoom.south;
  }

  public lookEast() {
    return this.currentRoom.east;
  }

  public lookWest() {
    return this.currentRoom.west;
  }

  public itemSearch() {
    const items = this.currentRoom.items;
    return items.length > 0 ? items : null;
  }

  public bagSearch() {
    const items = this.items;
    return items.length > 0 ? items : null;
  }
}
